bpo-41538: Make python interpreter customizable for EnvBuilder

abbec opened the pull request:
This adds a new keyword parameter `python_interpreter` to
`venv.EnvBuilder` that lets the user customize which python to use when
creating the virtualenv, much like the virtualenv command's `-p` flag.

The reason that this is needed is that for some embedded cases (Blender
being one example), `sys._base_executable` is not set to a valid Python
executable.

<!-- issue-number: [bpo-41538](https://bugs.python.org/issue41538) -->
https://bugs.python.org/issue41538
<!-- /issue-number -->

abbec committed:
bpo-41538: Make python interpreter customizable for EnvBuilder

This adds a new keyword parameter `python_interpreter` to
`venv.EnvBuilder` that lets the user customize which python to use when
creating the virtualenv, much like the virtualenv command's `-p` flag.

The reason that this is needed is that for some embedded cases (Blender
being one example), `sys._base_executable` is not set to a valid Python
executable.

vsajip reviewed (changes requested):
Not sure `python_interpreter` is the best name - no other sort of interpreter would be valid here, right? Perhaps you could use `executable` or `venv_executable` as the keyword argument and attribute name, as that's more commonly used to refer to a path to a Python interpreter executable (for example, `sys.executable` for the running interpreter).

Also, a test should be implemented for this (which would only be run if an alternative Python was available in the system path, and skipped otherwise).

abbec committed:
Rename `python_interpreter` to `executable`

abbec committed:
Add venv custom executable test case

This test case verifies that if you set a custom executable, it is
actually being used and the virtualenv can be created using it. Note
that this test is skipped if an alternative python can not be found in
path.

vsajip commented on Lib/test/test_venv.py:
```diff
@@ -33,6 +33,12 @@
     'cannot run venv.create from within a venv on this platform')
 
+requireAlternativePythonInPath = unittest.skipUnless(
+        shutil.which("python") is not None
```
This won't work if `python` resolves to a Python 2.x interpreter, as they don't know about Python's built-in venv (which appeared in Python 3.3). See comment below on how test should be structured.

vsajip commented on Lib/test/test_venv.py:
```diff
@@ -335,6 +341,27 @@ def test_executable_symlinks(self):
+
+        # make sure it can be created with the alternative executable
+        rmtree(self.env_dir)
+        self.run_with_capture(venv.create, self.env_dir)
```
This won't work because `venv.create()` won't call the builder you just created. Which makes me think:

1. `venv.create()` should also have the `executable` keyword argument, and pass it through to the builder that it creates.
2. The `main` function should parse an additional `--executable` argument to allow an executable to be passed in from the command line, and pass that to the builder it creates.

Also, I would suggest structuring the test as follows:

1. Remove the `requireAlternativePython` decorator and function.
2. Instead, do something along the lines of
```
    # create temp dir to hold venv
    found_a_python = False
    for minor in range(3, 11):
        executable = 'python3.%s' % minor
        alt_python = shutil.which(executable)
        if alt_python and alt_python != sys._base_executable:
            found_a_python = True
            # create builder with executable=alt_python
            # remove venv dir
            # create venv using the builder
            # check that venv creation appeared to work correctly
            # check that venv creation used the correct python interpreter by invoking it with
            # <venv_dir>/bin/python --version and checking what comes back is what's expected
    if not found_a_python:
        raise unittest.SkipTest('No alternative Python interpreters found for venv executable test')
```

abbec replied:
Yes, of course, will fix!

vsajip replied:
Possibly also add a "python3" just in case there's no "python3.X" available.

abbec committed:
Add `executable` argument to venv.create and CLI

Also, update the test.
